src/codegen/encoding/mod.rs

```diff
@@ -1732,10 +1735,57 @@ pub(super) trait AbiEncoding {
 
     /// Returns if the we are packed encoding
     fn is_packed(&self) -> bool;
+
+    /// Encode "Panic(uint256)" error data.
+    fn const_error_panic(&self, code: PanicCode) -> Vec<u8> {
+        let mut bytes = SolidityError::Panic(code).selector().to_be_bytes().to_vec();
+        bytes.push(code as u8);
+        bytes.resize(36, 0);
+        bytes
+    }
+
+    /// Encode "Error(string)" error data where the contents are compile time constant.
+    fn const_error_string(&self, data: String) -> Vec<u8> {
+        let mut bytes = SolidityError::String(Expression::Poison)
+            .selector()
+            .to_be_bytes()
+            .to_vec();
+        bytes.extend_from_slice(&data.encode());
+        bytes
+    }
+
+    /// Encode the error data at compile time.
+    ///
+    /// Returns`None` if the error data is not a compile time constant value.
+    fn encode_error_data_const(&self, error: SolidityError) -> Option<Expression> {
+        let bytes = match error {
+            SolidityError::Empty => return None,
+            SolidityError::String(data) => match data {
+                Expression::AllocDynamicBytes {
+                    ty: Type::String,
+                    initializer: Some(data),
+                    ..
+                } => self.const_error_string(String::from_utf8(data).unwrap()),
```

Comment:
There is no `Expression::StringLiteral` on `codegen::Expression`. In `emit::expression.rs`:

```
Expression::AllocDynamicBytes {
            ty,
            size,
            initializer,
            ..
        } => {
            if matches!(ty, Type::Slice(_)) {
                let init = initializer.as_ref().unwrap();

                let data = bin.emit_global_string("const_string", init, true)
```

This is exactly what I want. This only works by having an `AllocDynamicBytes`. `emit` even checks if it's a slice and then emits an interned string, which is exactly what I want

Reply:
You can create an `Expression::StringLiteral` that does this. This is what I meant.

Reply:
Ah I see. Yeah why not; I created #1453 so we don't forget doing it afterwards. For now I'd like to make progress with this PR, and creating a new Expression is out of scope here, especially if the same can be done by just using `AllocDynamicBytes`

Reply:
The new expression is the easiest solution for this PR, IMO, as it would not require the hack with `Expression::AllocaDynamicBytes`.

Reply:
It would be the proper solution, but still out of scope for this PR. On a side note, I don't see how implementing and testing a whole new Expression, and then replace existing parts where this new expression can also be used, is an easier Solution than just  using `Expression::AllocaDynamicByte`. It works as I expected with minimal code.

Reply:
`Expression::AllocDynamicBytes` gives you the constant string, but it is not the proper representation of it in the code. You are masking a constant string behind an allocation. We might implement passes in the future that discern between when you have a constant string and and you have slice (e.g. concatenating items during compile time) that simply won't work. The CFG should have a fairly trustworthy representation of the code we write.

Reply:
I can't recall right now any other place where string literals (or a byte array literal) are need. The slice we have in `Expression::AllocDynamicBytes` is needed due to Solana runtime calls.

Reply:
> I can't recall right now any other place where string literals (or a byte array literal) are need.


Obviously  is already happening at other places, otherwise the code turning `AllocDynamicBytes` of type `Slice` into interned strings wouldn't be there. 

> The slice we have in Expression::AllocDynamicBytes is needed due to Solana runtime calls.

Why is it ok then to use `AllocDynamicBytes` of type Slice when the Solana runtime needs it but not for other things. I don't get it.

> We might implement passes in the future that discern between when you have a constant string and and you have slice

Then we just fix it before implementing these optimizations. Also I'm not at all worried about this optimization not happening for this case.

This discussion did convince me 0% so far to do this. Again: Out of scope and I don't have time for this. I'm trying to fix real problems so we can get the compiler audited and supporting OpenZepplin as soon as possible. I lack any resource to improve things that otherwise work but are not perfect.

I see your point. I agree with you that this could be done different. However given no security or other significant issues arise from using `AllocDynamicBytes` here I simply can neither afford to scope creep this PR nor to implement such cosmetics before merging this.

Reply:
I re-reviewed your PR and re-analyzed the use-case. It makes sense to me now.